packages/widgets/src/blocks/legacy-widget/edit/preview.js

```diff
@@ -7,57 +7,102 @@ import classnames from 'classnames';
  * WordPress dependencies
  */
 import { useRefEffect } from '@wordpress/compose';
-import { addQueryArgs } from '@wordpress/url';
-import { useState } from '@wordpress/element';
+import { useState, useEffect } from '@wordpress/element';
 import { Placeholder, Spinner, Disabled } from '@wordpress/components';
 import { __ } from '@wordpress/i18n';
+import apiFetch from '@wordpress/api-fetch';
 
 export default function Preview( { idBase, instance, isVisible } ) {
 	const [ isLoaded, setIsLoaded ] = useState( false );
+	const [ srcDoc, setSrcDoc ] = useState( '' );
+	const [ isPreviewFetched, setIsPreviewFetched ] = useState( false );
+
+	const abortController =
+		typeof window.AbortController === 'undefined'
+			? undefined
+			: new window.AbortController();
+
+	async function fetchPreviewHTML() {
+		const restRoute = `/wp/v2/widget-types/${ idBase }/render`;
+		return await apiFetch( {
+			path: restRoute,
+			method: 'POST',
+			signal: abortController?.signal,
+			data: {
+				id_base: idBase,
+				instance,
+			},
+		} );
+	}
+
+	useEffect( () => {
+		if ( ! isPreviewFetched ) {
+			setIsPreviewFetched( true );
+			fetchPreviewHTML()
+				.then( ( response ) => {
+					setSrcDoc( response.preview );
+				} )
+				.catch( ( error ) => {
+					if ( error.name === 'AbortError' ) {
+						// We don't want to log abort errors.
+						return;
+					}
+					window.console.error(
+						`An error occurred while trying to fetch preview: ${ error.message }`,
+						error
+					);
+				} );
```

Comment:
I don't think we need to add custom error. We can just re-throw the error here.

Reply:
Yes, we don't have to add a custom error, but it will not hurt and give a user additional details about the error.

Reply:
It does increase the bundle size though, and the error message can't be minified. We could potentially use `@wordpress/warning` for this, it does some magic to exclude the error from prod.
I don't think throwing an error is that bad though, in fact, I think it's often time better than a simple `console.error` because it has stack traces available. And it's also a lot more visible so that we can monitor the unexpected early.

Reply:
> It does increase the bundle size though, and the error message can't be minified. 
Let's not use it then :) 
Fixed in a37caf465709f29a85b52fdd023e62681298712b